<?xml version="1.0"?>

<launch>

    <arg name="use_rviz" default="false" />

    <!-- ODOMETRY -->
    <!--rf2o_Laser_Odometry-->
    <include file="$(find art_racecar)/launch/includes/rf2o.launch.xml" />  

     <!-- Robot_Localization -->
    <node pkg="robot_localization" type="ekf_localization_node" name="ekf_se" clear_params="true">
        <rosparam command="load" file="$(find art_racecar)/param/ekf_params.yaml" />
    </node>

    <!-- gmapping -->
    <node pkg="gmapping" type="slam_gmapping" name="slam_gmapping" output="screen">
        <remap from="scan" to="scan"/>
        <param name="map_update_interval" value="2.5"/>
        <param name="maxUrange" value="16.0"/>
        <param name="sigma" value="0.05"/>
        <param name="kernelSize" value="1"/>
        <param name="lstep" value="0.05"/>
        <param name="astep" value="0.05"/>
        <param name="iterations" value="5"/>
        <param name="lsigma" value="0.075"/>
        <param name="ogain" value="3.0"/>
        <param name="lskip" value="0"/>
        <param name="srr" value="0.1"/>
        <param name="srt" value="0.2"/>
        <param name="str" value="0.1"/>
        <param name="stt" value="0.2"/>
        <param name="linearUpdate" value="0.10"/>
        <param name="angularUpdate" value="0.25"/>
        <param name="temporalUpdate" value="1.0"/>
        <param name="resampleThreshold" value="0.25"/>
        <param name="particles" value="30"/>
        <param name="xmin" value="-100.0"/>
        <param name="ymin" value="-100.0"/>
        <param name="xmax" value="100.0"/>
        <param name="ymax" value="100.0"/>
        <param name="delta" value="0.05"/>
        <param name="llsamplerange" value="0.01"/>
        <param name="llsamplestep" value="0.01"/>
        <param name="lasamplerange" value="0.005"/>
        <param name="lasamplestep" value="0.005"/>
        <param name="odom_frame" value="odom"/>
        <param name="base_frame" value="base_footprint"/>
    </node>


     <!-- Navstack -->
    <node pkg="move_base" type="move_base" respawn="false" name="move_base">
        <!-- local planner -->
         <!--
        <param name="base_global_planner" value="global_planner/GlobalPlanner"/>  
         -->
        <param name="base_global_planner" value="navfn/NavfnROS"/>  
        <param name="base_local_planner" value="dwa_local_planner/DWAPlannerROS"/>  
        

        <rosparam file="$(find art_racecar)/param/dwa_local_planner_params.yaml" command="load"/>
        <!-- costmap layers -->    
        <rosparam file="$(find art_racecar)/param/local_costmap_params.yaml" command="load"/>
        <rosparam file="$(find art_racecar)/param/global_costmap_params.yaml" command="load"/> 
        <!-- move_base params -->
        <rosparam file="$(find art_racecar)/param/base_global_planner_params.yaml" command="load"/>
        <rosparam file="$(find art_racecar)/param/move_base_params.yaml" command="load"/>
        <remap from="/odom" to="/odometry/filtered"/>

       
    </node>

    <!-- L1 controller -->  
    <node pkg="art_racecar" type="art_car_controller" respawn="false" name="art_car_controller" output="screen">       
        <!-- L1 -->
        <param name="Vcmd" value="0.8" /> <!-- speed of car m/s -->        
        <!-- ESC -->
        <param name="baseSpeed" value="150"/> <!-- pwm for motor constant speed, 1500: stop --> 
 

        <!-- Servo -->
        <param name="baseAngle" value="0"/> <!-- the middle pos of servo, for tt02: 87, for hsp: ? -->
        <param name="AngleGain" value="-80.0"/> <!-- for tt02: >0, for hsp: <0 -->	
        <param name="GasGain" value="10.0"/>
	<param name="p_turn" value="25.0"/>
	<param name="d_turn" value="8.0"/>
    <param name="dist" value="0.2"/>
	<param name="pid_speedProportion" value="0.0"/>
	<param name="pid_speedIntegral" value="0.0"/>
	<param name="pid_speedDerivative" value="0.0"/>

        <!-- remap from="/odometry/filtered" to="odom" / --> 
        <remap from="/move_base_node/NavfnROS/plan" to="/move_base/NavfnROS/plan" /> 
    </node>

    <!-- Rviz -->
    <node pkg="rviz" type="rviz" name="rviz" args="-d $(find art_racecar)/rviz/amcl.rviz" if="$(arg use_rviz)" />

</launch>

